docs/design/security_context.md

```diff
@@ -0,0 +1,158 @@
+2.  Implement a security context structure that is part of a service account. The
+    default context provider can then be used to apply a security context based
+    on the service account associated with the pod.
+    
+### Security Context Provider
+
+The Kubelet will have an interface that points to a `SecurityContextProvider`. The `SecurityContextProvider` is invoked before creating and running a given container:
+
+```go
+type SecurityContextProvider interface {
+    // ModifyContainerConfig is called before the Docker createContainer call.
+    // The security context provider can make changes to the Config with which
+    // the container is created.
+    // An error is returned if it's not possible to secure the container as 
+    // requested with a security context. 
+	ModifyContainerConfig(pod *api.BoundPod, container *api.Container, config *docker.Config) error
+	
+	// ModifyHostConfig is called before the Docker runContainer call.
+	// The security context provider can make changes to the HostConfig, affecting
+	// security options, whether the container is privileged, volume binds, etc.
+	// An error is returned if it's not possible to secure the container as requested 
+    // with a security context. 
+	ModifyHostConfig(pod *api.BoundPod, container *api.Container, hostConfig *docker.HostConfig)
+}
+```
+If the value of the SecurityContextProvider field on the Kubelet is nil, the kubelet will create and run the container as it does today.   
+
+### Security Context
+
+A security context has a 1:1 correspondence to a service account and it can be included as
+part of the service account resource. Following is an example of an initial implementation:
+
+```go
+type SecurityContext struct {
+    // user is the uid to use when running the container
+	User int
+	
+	// allowPrivileged indicates whether this context allows privileged mode containers
+	AllowPrivileged bool
+	
+	// allowedVolumeTypes lists the types of volumes that a container can bind
+	AllowedVolumeTypes []string
+	
+	// addCapabilities is the list of Linux kernel capabilities to add
+	AddCapabilities []string
+	
+	// removeCapabilities is the list of Linux kernel capabilities to remove
+	RemoveCapabilities []string
+	
+	// SELinux specific settings (optional)
+	SELinux *SELinuxContext
+	
+	// AppArmor specific settings (optional)
+	AppArmor *AppArmorContext
+	
+	// FUTURE:
+	// With Linux user namespace support, it should be possible to map
+	// a range of container uids/gids to arbitrary host uids/gids
+	// UserMappings []IDMapping
+	// GroupMappings []IDMapping
+}
+
+type SELinuxContext struct {
+    // MCS label/SELinux level to run the container under
+    Level string
+    
+    // SELinux type label for container processes
+    Type  string    
+    
+    // FUTURE:
+    // LabelVolumeMountsExclusive []Volume
+    // LabelVolumeMountsShared    []Volume
+}
+
+type AppArmorContext struct {
+	// AppArmor profile
+	Profile string
+}
```

Comment:
----- Original Message -----
> > +type SELinuxContext struct {
> > +    // MCS label/SELinux level to run the container under
> > +    Level string
> > +
> > +    // SELinux type label for container processes
> > +    Type  string
> > +
> > +    // FUTURE:
> > +    // LabelVolumeMountsExclusive []Volume
> > +    // LabelVolumeMountsShared    []Volume
> > +}
> > +
> > +type AppArmorContext struct {
> > +	// AppArmor profile
> > +	Profile string
> > +}
> 
> For example, if my intent is: "This container should not have the same
> identity as any other container, both for node-local resources and for
> shared (storage) resources", then I am pretty sure the system could
> automatically come up with a User, SELinux.Level, SELinux.Type, and
> AppArmor.Profile settings.    The question that needs some thought is
> whether most other intents can similarly be expressed at an abstract level.

This is an interesting question - volumes are very low level (give me EXACTLY this thing).  Security context as modeled is a bit more like volumes.  It means a finalizer goes and turns a generic intent (maybe expressed by the admin or the namespace) onto a specific context on the pod (pod should run as this UID).  Your suggesting the opposite, something that the user can set "hey, I want this kind of security context", and then something has to go finalize and specialize it.